File: Python3/reverse_string_II.py
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        def reverse(s, pos = 0, length = len(s)):
            s = list(s)
            final = pos + length - 1
            for i in range(length//2):
                tmp = s[pos+i]                
                s[pos+i] = s[final-i]
                s[final-i] = tmp
            return "".join(s)

        for i in range(0, len(s), 2*k):
            if len(s[i:]) < k: s = reverse(s, i, len(s[i:]))
            elif len(s[i:]) >= k and len(s[i:]) < 2*k: s = reverse(s, i, k)
            else:
                s = reverse(s, i, k)
        return s

File: C/reverse_string_II.c
void reverse(char* s, int pos, int length){
    int f = pos + length - 1;
    for(int i = 0; i < length/2; i++){
        char* tmp = s[pos+i];                
        s[pos+i] = s[f-i];
        s[f-i] = tmp;
    }
}

char * reverseStr(char * s, int k){
    for(int i = 0; i < strlen(s); i = i + 2*k){
        if(strlen(s)-i < k) reverse(s, i, strlen(s)-i);
        else if(strlen(s)-i >= k && strlen(s)-i < 2*k) reverse(s, i, k);
        else reverse(s, i, k);
    }
    return s;
}
